pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/AbstractMetadataStore.java

```diff
@@ -496,6 +496,17 @@ public void registerSessionListener(Consumer<SessionEvent> listener) {
 
     protected void receivedSessionEvent(SessionEvent event) {
         isConnected = event.isConnected();
+
+        // Clear cache after session expired.
```

Comment:
Would it make sense to do it after new session is re-established, so that we would be able to continue with what we have in cache.

Reply:
@merlimat 

1. Session expires
2. ZK nodes updated
  a. **(Highlight)**: notifications will be lost and makes the cache dirty.
3. Session re-established

Even though the session is re-established, it may lose some notifications, so clearing the cache is better.

Reply:
> Even though the session is re-established, it may lose some notifications, so clearing the cache is better.

Yes, I was just thinking to do the clearing, though at the time of reconnection, not when the session has expired.

Reply:
Good pointer, it has been fixed. Thanks

Reply:
> Good pointer, it has been fixed. Thanks

@poorbarcode how was @merlimat's comment addressed? It might be better to postpone the flushing of caches until reconnection has finished.

Reply:
@lhotari @merlimat 

Sorry, I missunderstood before, it has been fixed now